### PR TITLE
Block hash getter for EVM

### PR DESCRIPTION
**Pull request by JcBernack**
- added `HistoryBlockHashProvider` to `BlockContext`
- added callback mechanism for libevm, based on passing integer handles to libevm instead of actual function pointers
- allow `BLOCKHASH` opcode to invoke a callback performing the block hash lookup via the block hash provider on the block context

Todos
- [x] refactor selection of callback handles in `LibEvmCallback` (probably use a map instead of a fixed array)
- [x] add more unit tests
- [x] add python test using a contract using the `BLOCKHASH` opcode

### CODE REVIEW

**Review by saratnt — APPROVED** [review]

File: sdk/src/main/java/com/horizen/account/state/EvmMessageProcessor.java [hunk 1]
```diff
@@ -52,15 +60,32 @@ public byte[] process(Message msg, BaseAccountStateView view, GasPool gas, Block
...
+        private BlockHashGetter(HistoryBlockHashProvider provider) {
+            this.provider = provider;
+        }
+
+        @Override
+        protected Hash getBlockHash(BigInteger blockNumber) {
+            return OptionConverters
+                .toJava(provider.blockIdByHeight(blockNumber.intValueExact()))
+                .map(hex -> new Hash(BytesUtils.fromHexString(hex)))
+                .orElse(Hash.ZERO);
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply sdk/src/main/java/com/horizen/account/state/EvmMessageProcessor.java:88]: yes

File: libevm/lib/evm_test.go [hunk 2]
```diff
@@ -275,13 +276,30 @@ func TestEvmOpCodes(t *testing.T) {
...
 	}
 
+	// setup callback proxy for the BLOCKHASH opcode
+	const blockHashCallbackHandle = 5132
+	SetCallbackProxy(func(handle int, args string) string {
+		switch handle {
+		case blockHashCallbackHandle:
+			actual := new(big.Int)
+			actual.SetString(args[2:], 16)
+			if expected := new(big.Int).Sub(blockNumber, common.Big1); actual.Cmp(expected) != 0 {
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply libevm/lib/evm_test.go:294]: done

File: evm/src/main/java/com/horizen/evm/LibEvmCallback.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * Base class to be used when passing callbacks to libevm. Can and should be used when passing parameter objects via the
+ * LibEvm.invoke() JSON interface. When a parameter derived from this type is passed to libevm it will be serialized as
+ * an integer handle which maps to the instance of the callback. Callback handles need to be disposed when not in use
+ * anymore, call close() - or better - use the try-with-resources syntax.
+ */
+abstract class LibEvmCallback implements AutoCloseable {
+    // this singleton instance of the callback will be passed to libevm,
+    // the static reference here will also prevent the callback instance from being garbage collected,
+    // because without it the only reference might be from native code (libevm) and the JVM does not know about that
+    static MasterCallback callackHandler = new MasterCallback();
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply evm/src/main/java/com/horizen/evm/LibEvmCallback.java:19]: agree

File: evm/src/main/java/com/horizen/evm/LibEvmCallback.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    private static final Logger logger = LogManager.getLogger();
+
+    private static final int MAX_CALLBACKS = 10;
+    private static final LibEvmCallback[] callbacks = new LibEvmCallback[MAX_CALLBACKS];
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply evm/src/main/java/com/horizen/evm/LibEvmCallback.java]: done

File: libevm/lib/evm.go [hunk 5]
```diff
@@ -31,14 +32,27 @@ type EvmParams struct {
...
 }
 
+type BlockHashCallback struct{ Callback }
+
+func (c *BlockHashCallback) getBlockHash(blockNumber uint64) common.Hash {
+	blockNumberBig := new(big.Int).SetUint64(blockNumber)
+	if c == nil {
+		// fallback to mocked block hash
+		return common.BytesToHash(crypto.Keccak256([]byte(blockNumberBig.String())))
+	}
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply libevm/lib/evm.go:42]: The null case is very useful for tests. We often don't even specify an `EvmContext` at all. I think we should leave it for now.

**Review by i-Alex — COMMENTED** [review]
So far looks good!

File: evm/src/main/java/com/horizen/evm/LibEvmCallback.java [hunk 6]
```diff
@@ -16,43 +19,42 @@ abstract class LibEvmCallback implements AutoCloseable {
...
-    private static final LibEvmCallback[] callbacks = new LibEvmCallback[MAX_CALLBACKS];
+    private static final Map<Integer, LibEvmCallback> callbacks = new HashMap<>();
 
     private static synchronized int register(LibEvmCallback callback) {
-        for (int i = 0; i < callbacks.length; i++) {
-            if (callbacks[i] == null) {
-                callbacks[i] = callback;
-                logger.trace("registered callback with handle {}: {}", i, callback);
-                return i;
+        // note: with N items in the map this will iterate for N+1 times, hence it should always find an unused handle
```
[the remark this answers is not in the document]
Comment (JcBernack) [reply evm/src/main/java/com/horizen/evm/LibEvmCallback.java:29]: True, the number of simultaneously active callbacks should be pretty low at all times though.

**Review by i-Alex — APPROVED** [review]

**Pull request merged**